bot/extensions/adventofcode/commands.py

```diff
@@ -0,0 +1,222 @@
+            )
+
+    @app_commands.command()
+    async def unsubscribe(self, interaction: core.InteractionType):
+        """Unsubscribe from receiving notifications for new puzzles"""
+
+        if self.role in interaction.user.roles:
+            await interaction.user.remove_roles(self.role)
+            await interaction.response.send_message(
+                "Okay! You have been __unsubscribed__ from notifications about new Advent of Code puzzles.",
+                ephemeral=True,
+            )
+        else:
+            await interaction.response.send_message(
+                "Hey, you don't even get any notifications about new Advent of Code puzzles currently anyway.",
+                ephemeral=True,
+            )
+
+    @app_commands.command()
+    async def countdown(self, interaction: core.InteractionType):
+        """Get the time left until the next puzzle is released"""
+
+        now = datetime.now(tz=pytz.timezone("EST"))
+        if now.month == 12:
+            # If it's December, calculate time until the next midnight EST time
+            target = now.replace(hour=0, minute=0, second=0, microsecond=0) + timedelta(days=1)
+        else:
+            # If it's not December, calculate time until the first of December
+            target = datetime(now.year, 12, 1, tzinfo=pytz.timezone("EST"))
+
+        if target < now:
+            return await interaction.response.send_message(
+                "Advent of Code is over for this year! See you next year!", ephemeral=True
+            )
+
+        await interaction.response.send_message(
+            f"There are {discord.utils.format_dt(target, 'R')} left until the next puzzle is released!",
+            ephemeral=True,
+        )
+
+    @app_commands.command(name="join")
+    async def join_leaderboard(self, interaction: core.InteractionType):
+        """Learn how to join the leaderboard"""
+
+        await interaction.response.send_message(
+            "Head over to https://adventofcode.com/leaderboard/private"
+            f"with the code `{LEADERBOARD_CODE}` to join the TWT private leaderboard!",
+            ephemeral=True,
+        )
+
+    @app_commands.command()
+    async def leaderboard(self, interaction: core.InteractionType):
+        """Get a snapshot of the TWT private AoC leaderboard"""
+
+        if interaction.channel_id != settings.aoc.channel_id:
+            return await interaction.response.send_message(
+                f"Please use the <#{settings.aoc.channel_id}> channel", ephemeral=True
+            )
+
+        async with aiohttp.ClientSession(cookies=AOC_SESSION_COOKIE, headers=AOC_REQUEST_HEADER) as session:
+            async with session.get(API_URL) as resp:
+                if resp.status == 200:
+                    leaderboard = await resp.json()
+                else:
+                    resp.raise_for_status()
+
+        members = [Member(**member_data) for member_data in leaderboard["members"].values()]
```

Comment:
Maybe create a helper for this
```python
async def fetch_leaderboard(leaderboard_id: int = None, year: int = None) -> Leaderboard:
```

If leaderboard is not passed then fetch global leaderboard
If year is not passed then default to the current year

Reply:
ngl sounds like a fun command to make